### ISSUE TITLE
[BUG] HL-1110 "out of memory" error when sending print job in 1200dpi/HQ mode

### ISSUE THREAD

**Issue by jaseg** [Owl-Maintain/brlaser#13]
**Describe the bug**

When trying to print a page of A4 text using 1200 DPI resolution, the CUPS job completes, but the printer never prints anything. The printer's yellow warning LED blinks four times in a row with a one-second pause, and afterwards the printer's maintenance log (printed by pressing the power button three times with the printer turned on already) shows an "out of memory" error. The printer is super cheap and has like, a megabyte of memory, so that checks out if it gets send the whole page at once.

**Steps to Reproduce**

AFAICT just print any A4 page of text in 1200 DPI mode.

**Desktop:**
 - Operating System: Arch, kernel 6.7.3-arch1-2
 - brlaser Version: current master, commit bfaf936

**Comment by danwood76** [comment]
I have just run into this issue trying to print a PDF, lowing quality from 600 dpi -> 300 dpi made it print.

I then tried to print using the official brother linux drivers and these fail in the same way (lowering from Fine to Normal made it print).

For tits and pickles I then booted up my virtual windows and it failed to print there too, however this time it popped up a window saying "Memory Full" and then text saying "Change the Print Quality to "Draft" or "Normal" in the driver settings and print again."

![Image](https://github.com/user-attachments/assets/30493248-1b6b-4a6f-a42e-ae1c29d97ae6)

So I guess it is just a cheapo laser with limited RAM and there is no way to print large pages with it, even the manufacturer can't do it!

**Comment by jaseg** [comment]
@danwood76 Yeah, that's what I figured in the end too. I even partially reverse engineered the Windows driver and found out that it neither supports streaming jobs that are too large for the printer's meager memory. The issue here is that the printer will accept bulk requests on its USB bulk out endpoint even if its memory is full (instead of stalling the request), which means the driver has no feedback from the printer at all. The only "fix" I can see for this that does not involve reverse engineering the printer's firmware would be to maybe hard-code a limit on the driver side to avoid the whole round trip through the printer and back and just show an error right away.
